File: Softcoinp.Backend/Services/AuditService.cs
using System;
using System.Linq;
using System.Text.Json;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Http;
using Softcoinp.Backend.Models;

namespace Softcoinp.Backend.Services
{
    public class AuditService : IAuditService
    {
        private readonly AppDbContext _db;
        private readonly IHttpContextAccessor _httpContextAccessor;

        public AuditService(AppDbContext db, IHttpContextAccessor httpContextAccessor)
        {
            _db = db;
            _httpContextAccessor = httpContextAccessor;
        }

        public async Task LogAsync(string action, string entity, Guid? entityId = null, object? data = null)
        {
            try
            {
                var http = _httpContextAccessor.HttpContext;
                Guid? userId = null;
                if (http?.User?.Identity?.IsAuthenticated == true)
                {
                    var idClaim = http.User.FindFirst("id")?.Value;
                    if (Guid.TryParse(idClaim, out var g)) userId = g;
                }

                string? ip = http?.Connection?.RemoteIpAddress?.ToString();
                string? ua = http?.Request?.Headers["User-Agent"].FirstOrDefault();

                var audit = new AuditLog
                {
                    Id = Guid.NewGuid(),
                    UserId = userId,
                    Action = action ?? string.Empty,
                    Entity = entity,
                    EntityId = entityId,
                    Data = data is null ? null : JsonSerializer.Serialize(data),
                    IpAddress = ip,
                    UserAgent = ua,
                    CreatedAt = DateTime.UtcNow
                };

                _db.AuditLogs.Add(audit);
                await _db.SaveChangesAsync();
            }
            catch
            {
                // No hacer nada si falla el logging
                // Evitar que falle la operación principal por un error en auditoría
            }
        }
    }
}

File: frontend-softcoinp/src/services/auditService.ts
import { getApiResponse } from "./api";

export interface AuditLog {
  id: string;
  userId: string | null;
  userName: string | null;
  action: string;
  entity: string | null;
  entityId: string | null;
  data: string | null;
  ipAddress: string | null;
  userAgent: string | null;
  createdAt: string;
}

export interface AuditLogsResponse {
  total: number;
  page: number;
  pageSize: number;
  logs: AuditLog[];
}

export const auditService = {
  getLogs: async (params?: { 
    page?: number; 
    pageSize?: number;
    action?: string;
    desde?: string;
    hasta?: string;
  }): Promise<AuditLogsResponse> => {
    const query = new URLSearchParams();
    if (params?.page) query.append("page", params.page.toString());
    if (params?.pageSize) query.append("pageSize", params.pageSize.toString());
    if (params?.action) query.append("action", params.action);
    if (params?.desde) query.append("desde", params.desde);
    if (params?.hasta) query.append("hasta", params.hasta);
    
    const response = await getApiResponse<AuditLogsResponse>(`/auditlogs?${query.toString()}`);
    return response.data;
  }
};
